rfc-128-continuous-deployment.md

```diff
@@ -0,0 +1,238 @@
+
+- Check the results of [the smoke tests](https://deploy.integration.publishing.service.gov.uk/job/Smokey/).
+- Look for any Icinga alerts related to the app.
+- Check the [deployment dashboard for the app](https://docs.publishing.service.gov.uk/manual/deployment-dashboards.html).
+- Generally make sure the change is tested.
+
+We often defer to automated tests to tell us when something is broken, but it's unclear whether this is really justified: there is no minimum standard for testing of GOV.UK apps.
+
+## Proposal
+
+The obvious solution to the backlog of deployments is to make them automatic. This has already begun, as part of [a trial of Continuous Deployment](https://docs.google.com/document/d/18y2fLzqbHyTliThxuiyxKznDw4TZmPiQZAk-3rkgpBg/edit) and [a subsequent extension to include 5 more apps](https://docs.google.com/document/d/1YhgjJjDRB57-IlgPNAAZzFTI3jIUucGJKokti_lQiRQ/edit).
+
+> **It is still possible to deploy an app manually.** The new pipeline is just an automation of the manual deployment steps. Rollback is unchanged, and freezing deployments is supported [[1](https://github.com/alphagov/govuk-puppet/blob/c24ff191ce7fc8a38ceb464f0139faba04b9734b/modules/govuk_jenkins/templates/jobs/deploy_app_downstream.yaml.erb#L22)].
+
+However, we have been unable to make further progress due to the other two issues: we are concerned that enabling automatic deployments en masse would lead to an increase in defects reaching production.
+
+In order to proceed, we need to agree a clear and actionable set of criteria for enabling automatic deployments, so that we will only do so when we can show this is safe, as well as secure.
+
+### Security checks
+
+We need to ensure that only people with [production access](https://docs.publishing.service.gov.uk/manual/rules-for-getting-production-access.html) are able to deploy code to the live site. This is equivalent to being a member of the "GOV.UK Production" GitHub team. As part of the trial of CD, we restricted merge access on the affected repos to people in this team [[1](https://github.com/alphagov/govuk-saas-config/pull/49)] [[2](https://github.com/alphagov/govuk-saas-config/pull/51)] [[3](https://github.com/alphagov/govuk-saas-config/pull/55)]. We will [continue to use this approach](https://github.com/alphagov/govuk-saas-config/pull/52#issuecomment-651903746).
+
+As well as restricting who can merge, we also had to address a vulnerability with Integration Jenkins: someone without production access could manually trigger the CD pipeline by running the [`Deploy_App_Downstream` Jenkins job](https://github.com/alphagov/govuk-puppet/blob/c24ff191ce7fc8a38ceb464f0139faba04b9734b/modules/govuk_jenkins/templates/jobs/deploy_app_downstream.yaml.erb). In order to prevent this kind of unauthorised deployment, the job has additional checks:
+
+- The tag to deploy must be the highest `release_123`-type tag.
+- The tagged commit must match the latest commit on the master branch.
```

Comment:
```suggestion
As well as restricting who can merge, we also have restrictions around which version of the code can be deployed. Only commits with latest release tag on the deployment branch can be automatically deployed. This protects against anyone in the integration environment triggering a potentially malicious deployment thats is not latest release. The following checks in the [`Deploy_App_Downstream` Jenkins job](https://github.com/alphagov/govuk-puppet/blob/c24ff191ce7fc8a38ceb464f0139faba04b9734b/modules/govuk_jenkins/templates/jobs/deploy_app_downstream.yaml.erb) enforce this restriction:

- The tag to deploy must be the highest `release_123`-type tag.
- The tagged commit must match the latest commit on the master branch.

However, the pipeline could still be abused by:

- By manually altering the job definition. In this scenario, anyone with non production access could deploy any version of the code to the Staging environment. (but no further: only people with production access can access the job in Staging Jenkins [1], which also does the above checks.)

- By repeatedly invoking the Deploy_App_Downstream job with the latest release, flooding the production environment with deployments. This could impact our production environment in various ways including creating a chain of graceful restarts, loss of worker capacity, delays to non-malicious deployments etc.
```

Reply:
Thanks @theseanything! I've gone with slightly different wording, partly to fix the typos, but mainly to try and clarify the latter point as an accepted/able risk, instead of focussing on the consequences. Here's the commit: https://github.com/alphagov/govuk-rfcs/pull/128/commits/037ffeab428caf76a013d001e57bbcb3e8b4fa10.

I still think it's worth continuing a discussion of potential ways to mitigate this risk, and I'd like to use the original thread for that, to keep it simple. So, I'll resolve this one.